### PR TITLE
use hash map for IN equality operator between literal values

### PR DESCRIPTION
**Pull request by max-hoffman**

### CODE REVIEW

File: sql/expression/in.go [hunk 1]
```diff
@@ -142,3 +147,84 @@ func (in *InTuple) Children() []sql.Expression {
...
+func (hit *HashInTuple) String() string {
+	return fmt.Sprintf("(%s HASH IN %s)", hit.Left(), hit.Right())
+}
+
+func (hit *HashInTuple) DebugString() string {
+	return fmt.Sprintf("(%s HASH IN %s)", sql.DebugString(hit.Left()), sql.DebugString(hit.Right()))
+}
+
+func newInMap(expr sql.Expression) (map[interface{}]sql.Expression, bool, sql.Type, error) {
+	elements := make(map[interface{}]sql.Expression)
```
Comment (zachmu) [inline sql/expression/in.go:210]: Probably should manually hash everything in here and make this `map[uint64]sql.Expression`

File: enginetest/query_plans.go [hunk 2]
```diff
@@ -463,6 +463,13 @@ var PlanTests = []QueryPlanTest{
 			"         └─ IndexedTableAccess(mytable on [mytable.s])\n" +
 			"",
 	},
+	{
+		Query: `SELECT * FROM mytable WHERE i in (1, 2, 3, 4)`,
+		ExpectedPlan: "Filter(mytable.i HASH IN (1, 2, 3, 4))\n" +
+			" └─ Projected table access on [i s]\n" +
+			"     └─ IndexedTableAccess(mytable on [mytable.i])\n" +
```
Comment (zachmu) [inline enginetest/query_plans.go:470]: Something is weird here.

The tests above don't get `HASH IN` treatment. Why not?

The second thing is that when we push down the `IN` clause to the table as an index, we should in principle remove that predicate from the filter. But we aren't doing that.

Could be a followup change, but not having the filter at all is even better than having it fast, and should work for IN expressions.

File: sql/analyzer/apply_hash_in.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import (
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/expression"
+	"github.com/dolthub/go-mysql-server/sql/plan"
+)
+
+func applyHashIn(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope) (sql.Node, error) {
+	return plan.TransformUpCtx(n, nil, func(c plan.TransformContext) (sql.Node, error) {
+		if f, ok := c.Node.(*plan.Filter); ok {
+			if i, ok := f.Expression.(*expression.InTuple); ok {
```
Comment (zachmu) [inline sql/analyzer/apply_hash_in.go:26]: You need to operate on the entire expression tree. This fails for filters like 

`where i in (2,3) and s = 5`

File: sql/analyzer/apply_hash_in.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+package analyzer
+
+import (
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/expression"
+	"github.com/dolthub/go-mysql-server/sql/plan"
+)
+
+func applyHashIn(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope) (sql.Node, error) {
```
Comment (zachmu) [inline sql/analyzer/apply_hash_in.go:23]: Should write some unit tests for this

**Review by zachmu — APPROVED** [review]
Good start, needs a little cleanup.

I think this is directionally correct enough that you could do followups in a second PR.

File: enginetest/query_plans.go (hunk 2 above)
Comment (max-hoffman) [reply enginetest/query_plans.go:470]: the other `IN`s are converted to table joins. re: filter, i wonder if my special case is sidestepping other IN optimizations, which would be bad. i just wanted some way to verify that the transform is being applied correctly. analyzer unit tests would also accomplish that

Comment (max-hoffman) [reply enginetest/query_plans.go:470]: Yep my bad expression tree traversal was missing nested tuple ins. Made an issue for the filter/table access on the same primary key https://github.com/dolthub/go-mysql-server/issues/616.

File: sql/analyzer/apply_hash_in_test.go [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+						expression.NewLiteral(int64(1), sql.Int64),
+						expression.NewBindVar("foo_id"),
+					),
+				),
+				child,
+			),
+		},
+	}
+
+	for _, test := range tests {
```
Comment (zachmu) [inline sql/analyzer/apply_hash_in_test.go:148]: You should use `analyzerFnTestCase` and `runTestCases` for these

**Pull request merged**